File: ios/Classes/YandexPolygonController.swift
import YandexMapsMobile

class YandexPolygonController: NSObject, YandexMapObjectController {
  public let polygon: YMKPolygonMapObject
  private let tapListener: YandexMapObjectTapListener
  private unowned var controller: YandexMapController
  public let id: String

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let polygon = parent.addPolygon(with: Utils.polygonFromJson(params))

    self.polygon = polygon
    self.id = params["id"] as! String
    self.controller = controller
    self.tapListener = YandexMapObjectTapListener(id: id, controller: controller)

    super.init()

    polygon.userData = self.id
    polygon.addTapListener(with: tapListener)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    let style = params["style"] as! [String: Any]

    polygon.isGeodesic = (params["isGeodesic"] as! NSNumber).boolValue
    polygon.zIndex = (params["zIndex"] as! NSNumber).floatValue
    polygon.strokeColor = Utils.uiColor(fromInt: (style["strokeColor"] as! NSNumber).int64Value)
    polygon.strokeWidth = (style["strokeWidth"] as! NSNumber).floatValue
    polygon.fillColor = Utils.uiColor(fromInt: (style["fillColor"] as! NSNumber).int64Value)
    polygon.geometry = Utils.polygonFromJson(params)
    polygon.isVisible = (params["isVisible"] as! NSNumber).boolValue
  }

  public func remove() {
    polygon.parent.remove(with: polygon)
  }
}

File: android/src/main/java/com/unact/yandexmapkit/YandexPolygonController.java
package com.unact.yandexmapkit;

import androidx.annotation.NonNull;

import com.yandex.mapkit.geometry.Point;
import com.yandex.mapkit.map.MapObject;
import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.MapObjectTapListener;
import com.yandex.mapkit.map.PolygonMapObject;

import java.lang.ref.WeakReference;
import java.util.Map;

public class YandexPolygonController extends YandexMapObjectController implements MapObjectTapListener {
  public final PolygonMapObject polygon;
  private boolean consumeTapEvents = false;
  @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  private final WeakReference<YandexMapController> controller;
  public final String id;

  public YandexPolygonController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    PolygonMapObject polygon = parent.addPolygon(Utils.polygonFromJson(params));

    this.polygon = polygon;
    this.id = (String) params.get("id");
    this.controller = controller;

    polygon.setUserData(this.id);
    polygon.addTapListener(this);
    update(params);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void update(Map<String, Object> params) {
    polygon.setGeometry(Utils.polygonFromJson(params));
    polygon.setGeodesic((Boolean) params.get("isGeodesic"));
    polygon.setZIndex(((Double) params.get("zIndex")).floatValue());
    polygon.setVisible((Boolean) params.get("isVisible"));
    polygon.setStrokeWidth(((Double) params.get("strokeWidth")).floatValue());
    polygon.setStrokeColor(((Number) params.get("strokeColor")).intValue());
    polygon.setFillColor(((Number) params.get("fillColor")).intValue());

    consumeTapEvents = (Boolean) params.get("consumeTapEvents");
  }

  public void remove() {
    polygon.getParent().remove(polygon);
  }

  @Override
  public boolean onMapObjectTap(@NonNull MapObject mapObject, @NonNull Point point) {
    controller.get().mapObjectTap(id, point);

    return consumeTapEvents;
  }
}
